File: archive/python-prototype/tests/kernel_fixture.py
"""Controlled fixture for Kernel's opt-in real Memory regression; stdout is control only."""
import json
import os
from pathlib import Path
import sys
import threading
import time
sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from tekesmemory.cli import setup
from tekesmemory.common import canonical, private_json
from tekesmemory.server import Server

os.umask(0o077)
root = Path(sys.argv[1])
paths = setup(root / 'real-memory', 'ws', root)
config_path = Path(paths['service_config'])
config = private_json(config_path)
config['port'] = 0
config_path.write_text(canonical(config))
server = Server(config_path)
thread = threading.Thread(target=server.serve_forever, daemon=True)
thread.start()
adapter_path = Path(paths['adapter_config'])
adapter = private_json(adapter_path)
adapter['endpoint'] = f'http://127.0.0.1:{server.server_port}/mcp'
adapter['retrieval']['estimated_tokens'] = 8192
adapter_path.write_text(canonical(adapter))
principal = config['principals'][0]
scope = principal['scopes'][0]
server.store.call(principal, 'memory.save', {'schema_version': 1, 'scope': scope, 'kind': 'episode',
    'content': 'first second LIFECYCLE_CONTEXT_MARKER', 'sources': [{'host': 'fixture', 'ref': 'seed', 'digest': 'a'*64}], 'idempotency_key': 'seed'})
print(canonical({'hook_root': str(config_path.parent)}), flush=True)
try:
    line = sys.stdin.readline()
    if line.strip() == 'verify':
        for _ in range(100):
            with server.store.lock:
                records = [json.loads(r[0]) for r in server.store.db.execute("SELECT payload FROM observations WHERE state='completed'")]
            if any(r.get('type') == 'turn_outcome' for r in records): break
            time.sleep(.02)
        assert sum(r.get('type') == 'turn_outcome' for r in records) == 1
        assert sum(r.get('type') == 'tool_result' for r in records) == 1
        assert next(r for r in records if r.get('type') == 'tool_result')['payload']['outcome'] == 'ok'
        assert 'Still thinking' not in canonical(records)
        print(canonical({'verified': True, 'types': [r['type'] for r in records]}), flush=True)
finally:
    server.shutdown()
    thread.join(timeout=3)
    server.server_close()

File: examples/kernel_fixture.rs
//! Opt-in Kernel integration fixture: actual Rust service + extension, no Python.
use serde_json::{Value, json};
use std::{
    io::{BufRead, Write},
    path::PathBuf,
    process::{Command, Stdio},
    time::Duration,
};
use tekes_memory::{client::Client, common::*, config};
struct ChildGuard(std::process::Child);
impl Drop for ChildGuard {
    fn drop(&mut self) {
        let _ = self.0.kill();
        let _ = self.0.wait();
    }
}
fn main() {
    unsafe { libc::umask(0o077) };
    let root = PathBuf::from(std::env::args().nth(1).expect("root"));
    let binary = PathBuf::from(std::env::args().nth(2).expect("Rust service binary"));
    let paths = config::setup(&root.join("real-memory"), "ws", &root, 0).unwrap();
    let cp = PathBuf::from(paths["service_config"].as_str().unwrap());
    let ready = root.join("memory-ready.json");
    let mut service = ChildGuard(
        Command::new(&binary)
            .args(["serve", "--config"])
            .arg(&cp)
            .arg("--ready-file")
            .arg(&ready)
            .stdout(Stdio::null())
            .spawn()
            .unwrap(),
    );
    for _ in 0..200 {
        if ready.exists() {
            break;
        }
        assert!(service.0.try_wait().unwrap().is_none(), "service exited");
        std::thread::sleep(Duration::from_millis(10));
    }
    let endpoint = private_json(&ready).unwrap()["endpoint"]
        .as_str()
        .unwrap()
        .to_owned();
    let ap = PathBuf::from(paths["adapter_config"].as_str().unwrap());
    let mut adapter = private_json(&ap).unwrap();
    adapter["endpoint"] = json!(endpoint);
    adapter["retrieval"]["estimated_tokens"] = json!(8192);
    std::fs::write(&ap, canonical(&adapter)).unwrap();
    for e in std::fs::read_dir(cp.parent().unwrap().join("hooks")).unwrap() {
        let path = e.unwrap().path();
        let mut hook = private_json(&path).unwrap();
        hook["argv"] = json!([binary, "kernel", "--config", ap]);
        std::fs::write(path, format!("{}\n", canonical(&hook))).unwrap();
    }
    let c = private_json(&cp).unwrap();
    let scope = &c["principals"][0]["scopes"][0];
    let credential = private_json(&cp.parent().unwrap().join("model-credential.json")).unwrap();
    Client::open(&endpoint,text(&credential,"token").unwrap(),Duration::from_secs(5)).unwrap().call("memory.save",json!({"schema_version":1,"scope":scope,"kind":"episode","content":"first second LIFECYCLE_CONTEXT_MARKER","sources":[{"host":"fixture","ref":"seed","digest":"a".repeat(64)}],"idempotency_key":"seed"})).unwrap();
    println!("{}", canonical(&json!({"hook_root":cp.parent().unwrap()})));
    std::io::stdout().flush().unwrap();
    let mut line = String::new();
    std::io::stdin().lock().read_line(&mut line).unwrap();
    assert_eq!(line.trim(), "verify");
    let db = rusqlite::Connection::open(cp.parent().unwrap().join("data/memory.sqlite3")).unwrap();
    let mut records = vec![];
    for _ in 0..100 {
        let mut stmt = db
            .prepare("SELECT payload FROM observations WHERE state='completed'")
            .unwrap();
        records = stmt
            .query_map([], |r| r.get::<_, String>(0))
            .unwrap()
            .map(|r| serde_json::from_str::<Value>(&r.unwrap()).unwrap())
            .collect::<Vec<_>>();
        if records.iter().any(|r| r["type"] == "turn_outcome") {
            break;
        }
        std::thread::sleep(Duration::from_millis(20));
    }
    assert_eq!(
        records
            .iter()
            .filter(|r| r["type"] == "turn_outcome")
            .count(),
        1
    );
    assert_eq!(
        records
            .iter()
            .filter(|r| r["type"] == "tool_result")
            .count(),
        1
    );
    assert_eq!(
        records.iter().find(|r| r["type"] == "tool_result").unwrap()["payload"]["outcome"],
        "ok"
    );
    assert!(!canonical(&json!(records)).contains("Still thinking"));
    println!(
        "{}",
        canonical(
            &json!({"verified":true,"types":records.iter().map(|r|r["type"].clone()).collect::<Vec<_>>()})
        )
    );
    std::io::stdout().flush().unwrap();
}
